### ISSUE TITLE
nan(Dual{T}) has type T

### ISSUE THREAD

**Issue by oxinabox** [JuliaDiff/DualNumbers.jl#20]
nan(Dual{Float32})
     > NaN32

But a Dual NaN does exist, kinda:

    convert(Dual{Float32},NaN)
    >NaN32 + 0.0f0du

The definition of nan(f)

> Base.nan(f)
>
 >  Returns NaN (not-a-number) of the floating point type "f" or of
  >  the same floating point type as "f"

There are two options for what could be done.

Either `nan{Dual{T}}` should a NaN in the real part and zero in the du part -- it makes sense to match the behavior of convert, and for the nan operator to apply to the real part.

Or it should do as `nan(Complex{Float32})` does and throw an error:

    `nan` has no method matching nan(::Type{Complex{Float32}}) while loading In[1], in expression starting on line 1

My code assumes that `nan{T<:Number}` gives me a NaN of type T.
Or at very least makes it clear that it can't be done.

**Comment by mlubin** [comment]
How are you using ``nan`` in your code?

**Comment by oxinabox** [comment]
I'm not sure that this will help you,
but I use it when creating a lightweight version of one of my composite types.
(to save on network transmission, in my parallel code).
I want to set one of the fields to NaN so that if it is ever accesses (which on the lightweight version it shouldn't be), it becomes immediately apparent.
I further have a version of the lightweight type that support converting,
as part of me using it with ForwardDiff.jl

Yes there are other ways to do it, eg Nullable, which may be better.
But because I was using NaN as a quick hack,
I came across the aforementioned unexpected behavior.


```
type PVDM{N<:Number, S<:AbstractString}
    we::WE
    pe::WE #use a word embedder for Paragraphs too
    
    W::AbstractMatrix{N}
    b::AbstractVector{N}

    window_length::Int
    varience::N
end

function PVDM{N,S}(we::WE{N,S}, window_length::Int, varience=0.001)
    
    emb_width,n_words = size(we.L)
    concat_layer_width = emb_width*(window_length+1)
    const W = convert(Matrix{N}, varience*randn(n_words,concat_layer_width))
    const b = convert(Vector{N}, varience*randn(n_words))
    
    pe = WE(N,Vector{S},emb_width)
    
    PVDM{N,S}( we, pe, W,b, window_length, varience)
end

@doc "Lightwieght version, that does not have support for lookups or additions"
function PVDM_light{N,S}(pvdm::PVDM{N,S})
    we = WE_light(pvdm.we)
    pe = WE_light(pvdm.we)
    W = pvdm.W
    b = pvdm.b
    window_length = pvdm.window_length
    varience = nan(N)
    PVDM{N,S}( we, pe, W,b, window_length, varience)
end

@doc "Lightwieght version, that does not have support for lookups or additions, and has N converted type type N2"    
function PVDM_light{N,S}(pvdm::PVDM{N,S}, N2::DataType)
    we = WE_light(pvdm.we,N2)
    pe = WE_light(pvdm.we, N2)
    W = convert(Matrix{N2},pvdm.W)
    b = convert(Vector{N2},pvdm.b)
    window_length = pvdm.window_length
    varience = nan(N2)#Replace with convert(N2,NaN)
    PVDM{N2,S}( we, pe, W,b, window_length, varience)
end
    
```

**Comment by oxinabox** [comment]
It was added in the commit below, if that helps you remember why you made it that way.
I'm sure you had a good reason at the time.

https://github.com/JuliaDiff/DualNumbers.jl/commit/336b003af020e44e30064e9f1061f1a3b7de1a81

**Comment by mlubin** [comment]
Since we overload the comparison operators, I think it's safe to make this change. Will need some testing though.

**Comment by tlycken** [comment]
Now, `inf(T)` and `nan(T)` are deprecated in favor of `convert(T, Inf)` and `convert(T, NaN)`, so I think these should simply be removed (after ample deprecation).

**Comment by mlubin** [comment]
Resolved by #27
